compiler/src/pipeline.rs

```diff
@@ -0,0 +1,489 @@
+    parsed::{asm::ASMProgram, PILFile},
+    DiffMonitor,
+};
+use backend::{BackendType, Proof};
+use executor::{constant_evaluator, witgen::QueryCallback};
+use log::Level;
+use mktemp::Temp;
+use number::FieldElement;
+
+use crate::verify::{write_commits_to_fs, write_constants_to_fs, write_constraints_to_fs};
+
+pub struct GeneratedWitness<T: FieldElement> {
+    pub pil: Analyzed<T>,
+    pub constants: Vec<(String, Vec<T>)>,
+    pub witness: Option<Vec<(String, Vec<T>)>>,
+}
+
+pub struct PilWithConstants<T: FieldElement> {
+    pub pil: Analyzed<T>,
+    pub constants: Vec<(String, Vec<T>)>,
+}
+
+pub struct CompilationResult<T: FieldElement> {
+    /// Constant columns, potentially incomplete (if success is false)
+    pub constants: Vec<(String, Vec<T>)>,
+    /// Witness columns, potentially None (if success is false)
+    pub witness: Option<Vec<(String, Vec<T>)>>,
+    /// Proof, potentially None (if success is false)
+    pub proof: Option<Proof>,
+    /// Serialized low level constraints, potentially None (if success is false)
+    pub constraints_serialization: Option<String>,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum Stage {
+    AsmFile,
+    AsmString,
+    Parsed,
+    Resolved,
+    AnalyzedAsm,
+    PilConstraints,
+    Graph,
+    Linked,
+    PilFile,
+    PilString,
+    AnalyzedPil,
+    OptimizedPil,
+    PilWithConstants,
+    GeneratedWitness,
+    Proof,
+}
+
+enum Artifact<T: FieldElement> {
+    AsmFile(PathBuf),
+    AsmString(Option<PathBuf>, String),
+    Parsed(Option<PathBuf>, ASMProgram<T>),
+    Resolved(ASMProgram<T>),
+    AnalyzedAsm(AnalysisASMFile<T>),
+    PilConstraints(AnalysisASMFile<T>),
+    Graph(PILGraph<T>),
+    Linked(PILFile<T>),
+    PilFile(PathBuf),
+    PilString(String),
+    AnaylyzedPil(Analyzed<T>),
+    OptimzedPil(Analyzed<T>),
+    PilWithConstants(PilWithConstants<T>),
+    GeneratedWitness(GeneratedWitness<T>),
+    Proof(CompilationResult<T>),
+}
+
+pub struct Pipeline<T: FieldElement> {
+    /// The current artifact. It is never None in practice, making it an Option is
+    /// only necessary so that we can take ownership of it in advance().
+    artifact: Option<Artifact<T>>,
+    /// The diff monitor is used to track changes between pipeline stages.
+    diff_monitor: DiffMonitor,
+    /// Output directory for intermediate files. If None, no files are written.
+    output_dir: Option<PathBuf>,
+    /// The name of the pipeline. Used to name output files.
+    name: Option<String>,
+    /// Whether to overwrite existing files. If false, an error is returned if a file
+    /// already exists.
+    force_overwrite: bool,
+    /// The log level to use for this pipeline.
+    log_level: Level,
+    // The output directory if `Pipeline::with_tmp_output` was called.
+    // Note that there is some redundancy with `output_dir`, but the Temp
+    // object has to live for the lifetime of the pipeline, so we keep it here.
+    tmp_dir: Option<Temp>,
```

Comment:
Why is `output_dir` not enough?

Reply:
can't we just take a `Path` and the outside needs to make sure it lives long enough? Though I can see it's convenient to have it here. Fine with keeping